### PR TITLE
chore: Pairing with contacts complete

### PR DESCRIPTION
**Pull request by GabrielGircenko**
# What kind of change does this PR introduce?
 
Pairing with contacts:
- [x] Gus should go through comments "NOTE to Gus" first
- [x] Gabriel fix so that it works
- [ ] Gus update backend and verify that this PR works

### CODE REVIEW

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt [hunk 1]
```diff
@@ -62,6 +83,187 @@ class GatewayRepository @Inject constructor(
...
+//        )
+
+        return firstPartyEndpoint.authorizeIndefinitely(
+            match.receiverEndpointPublicKey,
+        )
+    }
+
+    // TODO Potentially delete the following function
+    // NOTE to Gus: In this implementation this function is not needed.
+    // Only if you tell me that the import has to be done for Gateway to function properly
```
Comment (gnarea) [inline app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt:253]: The gateway has nothing to do with this, but yeah, we absolutely need this code or else we won't be able to send messages to the contact.

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt [hunk 2]
```diff
@@ -62,6 +83,187 @@ class GatewayRepository @Inject constructor(
...
+//    private suspend fun importPairingAuth(auth: ByteArray) {
+//        // For the following line, I'm getting Cannot access class 'tech.relaycorp.relaynet.SessionKey'. Check your module classpath for missing or conflicting dependencies
+//        val contactEndpoint = PrivateThirdPartyEndpoint.import(auth)
+//
+//        // TODO DELETE THIS COMMENTED CODE
+//        // NOTE to Gus: All this logic is done after in the AccountRepository
```
Comment (gnarea) [inline app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt:259]: I couldn't find something equivalent in there. Do you really mean `AccountRepository`?

**Review by gnarea — CHANGES_REQUESTED** [review]

File: app/src/main/java/tech/relaycorp/letro/repository/ContactRepository.kt [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+                contactDao.update(contactToUpdate.copy(status = PairingStatus.RequestSent))
+            }
+        }
+
+        databaseScope.launch {
+            gatewayRepository.pairingMatchReceived.collect { dataModel: PairingMatchDataModel ->
+                val contactToUpdate = getContactFromDatabase(
+                    accountAddress = dataModel.receiverVeraId, // TODO DELETE NOTE to Gus: Notice the switch here. Is this correct?
+                    contactAddress = dataModel.senderVeraId,
```
[the remark this answers is not in the document]
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/ContactRepository.kt]: Addressed and renamed to "requester" and "contact"

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt [hunk 4]
```diff
@@ -62,6 +83,187 @@ class GatewayRepository @Inject constructor(
...
+            receiverEndpointPublicKey = Base64.decode(parts[3], Base64.NO_WRAP),
+        )
+    }
+
+    private suspend fun generatePairingAuth(
+        match: PairingMatchDataModel,
+        firstPartyEndpoint: FirstPartyEndpoint,
+    ): ByteArray {
+        // TODO DELETE THIS COMMENTED CODE
+        // NOTE to Gus: All this logic is done before in the AccountRepository
```
[the remark this answers is not in the document]
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt:230]: ContactRepository* @gnarea

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt (hunk 1 above)
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt:253]: Is this comment obsolete? @gnarea Please, resolve it if yes

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt [hunk 5]
```diff
@@ -62,6 +83,187 @@ class GatewayRepository @Inject constructor(
...
+        return firstPartyEndpoint.authorizeIndefinitely(
+            match.receiverEndpointPublicKey,
+        )
+    }
+
+    // TODO Potentially delete the following function
+    // NOTE to Gus: In this implementation this function is not needed.
+    // Only if you tell me that the import has to be done for Gateway to function properly
+//    private suspend fun importPairingAuth(auth: ByteArray) {
+//        // For the following line, I'm getting Cannot access class 'tech.relaycorp.relaynet.SessionKey'. Check your module classpath for missing or conflicting dependencies
```
[the remark this answers is not in the document]
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt:255]: @gnarea can I delete this commented code?

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt (hunk 2 above)
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt]: @gnarea ContactRepository*

File: app/src/main/java/tech/relaycorp/letro/data/entity/ContactDataModel.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    indices = [Index("accountId")],
+)
+data class ContactDataModel(
+    @PrimaryKey(autoGenerate = true)
+    val id: Long = 0L,
+    val accountId: Long,
+    val address: String,
+    val alias: String,
+    val contactEndpointId: String? = null,
+    val contactEndpointPublicKey: ByteArray? = null, // TODO NOTE to Gus: Do we have to store this?
```
[the remark this answers is not in the document]
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/data/entity/ContactDataModel.kt]: Removed

File: app/src/main/java/tech/relaycorp/letro/repository/ContactRepository.kt [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+        databaseScope.launch {
+            gatewayRepository.pairingAuthorizationReceived.collect { dataModel: PairingMatchDataModel ->
+                val contactToUpdate = getContactFromDatabase(
+                    accountAddress = dataModel.receiverVeraId, // TODO DELETE NOTE to Gus: Notice the switch here. Is this correct?
+                    contactAddress = dataModel.senderVeraId,
+                ) ?: return@collect
+
+                contactDao.update(contactToUpdate.copy(status = PairingStatus.Complete))
+            }
```
[the remark this answers is not in the document]
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/ContactRepository.kt]: @gnarea let's jump on a call for this one

[the remark this answers is not in the document]
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/ContactRepository.kt]: Updated

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt]: Discussed and deleted

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt (hunk 2 above)
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt]: Discussed and resolved

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt (hunk 1 above)
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt]: Resolved

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (GabrielGircenko) [reply app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt]: Resolved

File: app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt [hunk 8]
```diff
@@ -62,6 +84,163 @@ class GatewayRepository @Inject constructor(
...
+                        _accountCreatedConfirmationReceivedFromServer.emit(
+                            AccountCreatedDataModel(
+                                veraIds[0],
+                                veraIds[1],
+                            ),
+                        )
+                        message.ack()
+                    }
+                    ContentType.ContactPairingMatch.value -> {
+                        // TODO Simplify this by moving to the ContactRepository
```
Comment (GabrielGircenko) [inline app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt:182]: @gnarea I thought about this while running through changes again. I would still keep this slightly more complicated writing for 2 reasons:
- follows the same principle for every message
- GatewayRepository is the only one "working" with Awala SDK

If you still think it should be simplified, then I would do the change for every message received. Please, let me know your thoughts

**Pull request closed without merging**